### ISSUE TITLE
Undocumented frontend build requirements

### ISSUE THREAD

**Issue by Vaelatern** [grafana/grafana#9861]
Froma  git source checkout for Grafana-4.6.1, trying to package it for Void Linux.

When trying to build frontend source, npm fails miserably. I expect the build to be possible as documented, but instead I got a lot of unexpected issues.

python2 is an undocumented dependency: https://gist.github.com/8b7fb2446bf07c8129dbc1ed3d5d78a2
Node sass cares too much about how it runs: https://gist.github.com/anonymous/80c1e00594906405d8003b9c9217b5fb
Building libsass fails miserably: https://gist.github.com/b07855cab6687624e4c756ed937e1716

Proposed solution: Check in built artifacts so no calls to `npm` are necessary, since the frontend does not contain binary artifacts.

(cc @the-maldridge)

**Comment by torkelo** [comment]
requiring npm install is very standard for all systems having a modern frontend. Checking in node_modules would add about 300mb of bloat to the repo. 

nodejs is a documented dependency to build grafana. And doing a npm or yarn install build dependencies has so far never been an issue. 

Maybe there is an open issue with Void Linux and node sass?

**Comment by Vaelatern** [comment]
There is not. I'm not suggesting including node_modules, simply adding the compiled artifacts as served by the frontend.

The fact that python2 slipped through the dependency list alone indicates that builds are less portable than tested.

**Comment by daniellee** [comment]
@Vaelatern the python2 dependency is usually only a problem on Windows (which is documented in the Windows section). Node-gyp works differently on different platforms.

> that builds are less portable than tested.

Can you explain more about why this is important to you? Are you talking about repeatable builds?

The build works on Windows, Mac OS and most Linux builds. We have never tested building on Void Linux so I'm not sure why you expect the instructions to be complete for that?

**Comment by the-maldridge** [comment]
I'll step in here as I'm also working on the Void package of grafana.  I don't think we have any expectation of the build instructions being specific to Void or any other distro for that matter.  What is the expectation is that they work at all (which, as has now been explained they do not).

The python dependency is very much a firm dependency as far as we have observed, even on Void Linux.  If the software quality of node-gyp is so low that it can't adequately identify what OS its running on, then that would explain a lot.

What has been suggested here is to check in the things that result from the node build phase, i.e. the 'compiled' javascript assets.  As with most "modern frontend" systems, we've found that most projects check in these artifacts specifically for this reason -- building them is not nearly as portable as it should be.  If this would introduce several hundred megabytes of JavaScript blobs, please say so.  I suspect thought that the earlier "300mb" figure is from the dependencies brought in by yarn.

TL;DR important because as far as we're concerned right now, Grafana doesn't build.

**Comment by daniellee** [comment]
@the-maldridge pretty much every linux distro has python2. I don't really understand what the problem is as it is only a build-time dependency. Why can you not install python2 on Void Linux? 

The docs take for granted that a developer has a working npm and node (which I do not think is unreasonable in 2017). We can document that node-gyp requires python2 if that is the issue here?

Checking in the compiled JS assets would be very painful for us that work on Grafana every day and I'm not seeing anything in your answer that would motivate us to drastically change our frontend workflow.

**Comment by the-maldridge** [comment]
@daniellee I can install python2 and am happy to do so, what I am taking issue with is that it is not listed in the documentation as a build time dependency on Linux but only on Windows.  It makes me wonder what else is missing from the documentation.

We have provisionally isolated the nodejs build issue to problems with node and PIE builds, so that will continue to be debugged as a Void specific issue on our side.

I will say though I am frustrated by the insta-close ticket response to this.  An issue was raised, its been repeated now several times that python2 should be added to the build dep list, and yet there is still some question that a documentation fault exists?  I'll accept that the node ecosystem is not designed to be maintainable and versioned in the same way Golang is (and in fact breaks the normal go build flow), but I still have the expectation that the project readme is complete.

As this is a closed bug I'm dropping this here.

**Comment by torkelo** [comment]
We can mention python dependency list  if you want, just so many use Linux systems that ship with it or is automatically installed when you install nodejs. This is the first time in 3 years and hundreds of contributors that someone has had an issue with node using python (on Linux at least)
